src/validators/validators.ts

```diff
@@ -7,16 +7,16 @@
 
 import { IFormContext } from '../components/FormContext';
 import { TBasicFieldValue } from '../components/withField';
-import { FieldErrorMessageId, TFieldError, TValidator } from './validators.types';
+import { FieldErrorMessageId, TAsyncValidator, TFieldError, TValidator } from './validators.types';
 
 /**
  * Wrapper function to call validators with parameters
  * @param validator function to call
  * @param context form context
  * @param args parameters for the validator
  */
-const withParam = (validator: TValidator, ...args: unknown[]): TValidator => {
-  return (value: TBasicFieldValue, context: IFormContext): TFieldError => validator(value, context, args);
+const withParam = <TCallback extends TValidator | TAsyncValidator>(validator: TCallback, ...args: unknown[]): TCallback => {
+  return (value: TBasicFieldValue, context: IFormContext): ReturnType<TCallback> => validator(value, context, args);
```

Comment:
@christophka In raw javascript the withParam method can already deal with asyncValidators. It's only the type checking that makes issues. I've tried to make withParam generic, but typescript complains about my return type. Can you take a look at that?

Reply:
So the problem seems to be the union type for TCallback regarding the sync / async. Also the build-in type `ReturnType` also seems to fail with this kind of setup.

So i changed it, so that we have the union type only for the return value for the validator. so instead of `<TCallback extends TValidator | TAsyncValidator>` I introduced a new type that returns a promise or not. that seems to fix the problem. Also the return type then can be infered correctly